spyder/api/plugins.py

```diff
@@ -8,69 +8,165 @@
+
+    # Define configuration defaults if using a separate file.
+    # List of tuples, with the first item in the tuple being the section
+    # name and the second item being the default options dictionary.
+    #
+    # CONF_DEFAULTS_EXAMPLE = [
+    #     ('section-name', {'option-1': 'some-value',
+    #                       'option-2': True,}),
+    #     ('another-section-name', {'option-3': 'some-other-value',
+    #                               'option-4': [1, 2, 3],}),
+    # ]
+    CONF_DEFAULTS = None
 
-    def __init__(self, parent=None):
-        super(BasePlugin, self).__init__(parent)
+    # Define configuration version if using a separate file
+    #
+    # IMPORTANT NOTES:
+    # 1. If you want to *change* the default value of a current option, you
+    #    need to do a MINOR update in config version, e.g. from 3.0.0 to 3.1.0
+    # 2. If you want to *remove* options that are no longer needed or if you
+    #    want to *rename* options, then you need to do a MAJOR update in
+    #    version, e.g. from 3.0.0 to 4.0.0
+    # 3. You don't need to touch this value if you're just adding a new option
+    CONF_VERSION = None
 
-        # This is the plugin parent, which corresponds to the main
-        # window.
+    # --- Private methods ----------------------------------------------------
+    # ------------------------------------------------------------------------
+    def _initialize(self, parent, configuration):
         self.main = parent
+        self._config = configuration or CONF
+
+        # Filesystem path to the root directory that contains the plugin.
+        self.PLUGIN_PATH = os.path.dirname(inspect.getfile(self.__class__))
 
-        # Filesystem path to the root directory that contains the
-        # plugin
-        self.PLUGIN_PATH = self._get_plugin_path()
+        # Register plugin configuration
+        if self._config is not None:
+            self._config.register_plugin(self)
+
+        # Check plugin compatibility
+        check_compatibility, message = self.check_compatibility()
+        self._is_compatible = True
+        if not check_compatibility:
+            self._is_compatible = False
+            self._show_compatibility_message(message)
 
         # Connect signals to slots.
         self.sig_show_status_message.connect(self.show_status_message)
         self.sig_option_changed.connect(self.set_option)
 
-    @Slot(str)
-    @Slot(str, int)
-    def show_status_message(self, message, timeout=0):
+    def _show_compatibility_message(self, message):
+        """Show a compatibility message."""
+        messageBox = QMessageBox(self)
+        messageBox.setWindowModality(Qt.NonModal)
+        messageBox.setAttribute(Qt.WA_DeleteOnClose)
+        messageBox.setWindowTitle(_('Compatibility Check'))
+        messageBox.setText(message)
+        messageBox.setStandardButtons(QMessageBox.Ok)
+        messageBox.show()
+
+    # TODO: This should be public? since it is used on mainwindow.py
```

Comment:
If it's not used by spyder-notebook, spyder-terminal, etc, directly, we should leave it as private.

Reply:
This was an old comment I forgot to erase ;-)